File: wrappers/LANDTRENDRWrapper.js
/**** Start of imports. If edited, may not auto-convert in the playground. ****/
var geometry = /* color: #d63000 */ee.Geometry.Polygon(
        [[[-121.91160307462638, 40.663194047718676],
          [-121.91434965665763, 40.4754220259598],
          [-121.54768095548576, 40.48273411595946],
          [-121.54630766447013, 40.66944403962391]]]);
/***** End of imports. If edited, may not auto-convert in the playground. *****/
//Wrapper for running z-score and linear trend across a moving window of years

//Module imports
var getImageLib = require('users/USFS_GTAC/modules:getImagesLib.js');
var dLib = require('users/USFS_GTAC/modules:changeDetectionLib.js');
///////////////////////////////////////////////////////////////////////////////
///////////////////////////////////////////////////////////////////////////////
dLib.getExistingChangeData();
// Define user parameters:

// 1. Specify study area: Study area
// Can specify a country, provide a fusion table  or asset table (must add 
// .geometry() after it), or draw a polygon and make studyArea = drawnPolygon
var studyArea =geometry;

// 2. Update the startJulian and endJulian variables to indicate your seasonal 
// constraints. This supports wrapping for tropics and southern hemisphere.
// startJulian: Starting Julian date 
// endJulian: Ending Julian date
var startJulian = 190;
var endJulian = 250

// 3. Specify start and end years for all analyses
// More than a 3 year span should be provided for time series methods to work 
// well. If using Fmask as the cloud/cloud shadow masking method, this does not 
// matter
var startYear = 1984;
var endYear = 2018;


// 4. Specify an annual buffer to include imagery from the same season 
// timeframe from the prior and following year. timeBuffer = 1 will result 
// in a 3 year moving window
var timebuffer = 1;

// 5. Specify the weights to be used for the moving window created by timeBuffer
//For example- if timeBuffer is 1, that is a 3 year moving window
//If the center year is 2000, then the years are 1999,2000, and 2001
//In order to overweight the center year, you could specify the weights as
//[1,5,1] which would duplicate the center year 5 times and increase its weight for
//the compositing method
var weights = [1,5,1];


// 6. Choose medoid or median compositing method. 
// Median tends to be smoother, while medoid retains 
// single date of observation across all bands
// If not exporting indices with composites to save space, medoid should be used
var compositingMethod = 'medoid';


// 7. Choose Top of Atmospheric (TOA) or Surface Reflectance (SR) 
// Specify TOA or SR
// Current implementation does not support Fmask for TOA
var toaOrSR = 'SR';

// 8. Choose whether to include Landat 7
// Generally only included when data are limited
var includeSLCOffL7 = false;

//9. Whether to defringe L5
//Landsat 5 data has fringes on the edges that can introduce anomalies into 
//the analysis.  This method removes them, but is somewhat computationally expensive
var defringeL5 = false;

// 10. Choose cloud/cloud shadow masking method
// Choices are a series of booleans for cloudScore, TDOM, and elements of Fmask
//Fmask masking options will run fastest since they're precomputed
//CloudScore runs pretty quickly, but does look at the time series to find areas that 
//always have a high cloudScore to reduce comission errors- this takes some time
//and needs a longer time series (>5 years or so)
//TDOM also looks at the time series and will need a longer time series
var applyCloudScore = false;
var applyFmaskCloudMask = true;

var applyTDOM = false;
var applyFmaskCloudShadowMask = true;

var applyFmaskSnowMask = true;

// 11. Cloud and cloud shadow masking parameters.
// If cloudScoreTDOM is chosen
// cloudScoreThresh: If using the cloudScoreTDOMShift method-Threshold for cloud 
//    masking (lower number masks more clouds.  Between 10 and 30 generally 
//    works best)
var cloudScoreThresh = 20;

// Percentile of cloud score to pull from time series to represent a minimum for 
// the cloud score over time for a given pixel. Reduces comission errors over 
// cool bright surfaces. Generally between 5 and 10 works well. 0 generally is a
// bit noisy
var cloudScorePctl = 10; 

// zScoreThresh: Threshold for cloud shadow masking- lower number masks out 
//    less. Between -0.8 and -1.2 generally works well
var zScoreThresh = -1;

// shadowSumThresh: Sum of IR bands to include as shadows within TDOM and the 
//    shadow shift method (lower number masks out less)
var shadowSumThresh = 0.35;

// contractPixels: The radius of the number of pixels to contract (negative 
//    buffer) clouds and cloud shadows by. Intended to eliminate smaller cloud 
//    patches that are likely errors
// (1.5 results in a -1 pixel buffer)(0.5 results in a -0 pixel buffer)
// (1.5 or 2.5 generally is sufficient)
var contractPixels = 1.5; 

// dilatePixels: The radius of the number of pixels to dilate (buffer) clouds 
//    and cloud shadows by. Intended to include edges of clouds/cloud shadows 
//    that are often missed
// (1.5 results in a 1 pixel buffer)(0.5 results in a 0 pixel buffer)
// (2.5 or 3.5 generally is sufficient)
var dilatePixels = 2.5;

// 12. correctIllumination: Choose if you want to correct the illumination using
// Sun-Canopy-Sensor+C correction. Additionally, choose the scale at which the
// correction is calculated in meters.
var correctIllumination = false;
var correctScale = 250;//Choose a scale to reduce on- 250 generally works well

//13. Export params
//Whether to export composites
var exportComposites = false;

//Set up Names for the export
var outputName = 'LT_';

//Provide location composites will be exported to
//This should be an asset folder, or more ideally, an asset imageCollection
var exportPathRoot = 'users/iwhousman/test/ChangeCollection';

// var exportPathRoot = 'projects/USFS/LCMS-NFS/R4/BT/Base-Learners/Base-Learners-Collection';
//CRS- must be provided.  
//Common crs codes: Web mercator is EPSG:4326, USGS Albers is EPSG:5070, 
//WGS84 UTM N hemisphere is EPSG:326+ zone number (zone 12 N would be EPSG:32612) and S hemisphere is EPSG:327+ zone number
var crs = 'EPSG:5070';

//Specify transform if scale is null and snapping to known grid is needed
var transform = [30,0,-2361915.0,0,-30,3177735.0];

//Specify scale if transform is null
var scale = null;
//List of bands or indices to iterate across
//Typically a list of spectral bands or computed indices
//Can include: 'blue','green','red','nir','swir1','swir2'
//'NBR','NDVI','wetness','greenness','brightness','tcAngleBG'
// var indexList = ee.List(['nir','swir1']);
var indexList = ['NBR','blue','green','red','nir','swir1','swir2','NDMI','NDVI','wetness','greenness','brightness','tcAngleBG'];
var indexListString = getImageLib.listToString(indexList,'_');

//The corresponding direction of forest loss for the given band/index specified above in indexList
// var ltDirection = [-1,    1];
var ltDirection =[1,-1,1,-1,    1,      1,   -1, -1,    -1,   -1,           -1,        1,          -1];


//Define landtrendr params
var run_params = { 
  maxSegments:            6,
  spikeThreshold:         0.9,
  vertexCountOvershoot:   3,
  preventOneYearRecovery: true,
  recoveryThreshold:      0.25,
  pvalThreshold:          0.05,
  bestModelProportion:    0.75,
  minObservationsNeeded:  6
};

//Define disturbance mapping filter parameters 
var treeLoss1  = 0.1;      //0.15 works well delta filter for 1 year duration disturbance, <= will not be included as disturbance - units are in units of segIndex defined in the following function definition
var treeLoss20 = 0.2;      //0.2 works well delta filter for 20 year duration disturbance, <= will not be included as disturbance - units are in units of segIndex defined in the following function definition
var preVal     = 0.01;      //0.2 works well. Set close to 0 if all pixels are wanted pre-disturbance value threshold - values below the provided threshold will exclude disturbance for those pixels - units are in units of segIndex defined in the following function definition
var mmu        = 0;       // minimum mapping unit for disturbance patches - units of pixels

var distParams = {
    tree_loss1: treeLoss1,
    tree_loss20: treeLoss20,  
    pre_val: preVal           
  };
  

//End params
///////////////////////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////
//Call on master wrapper function to get Landat scenes and composites
var lsAndTs = getImageLib.getLandsatWrapper(studyArea,startYear,endYear,startJulian,endJulian,
  timebuffer,weights,compositingMethod,
  toaOrSR,includeSLCOffL7,defringeL5,applyCloudScore,applyFmaskCloudMask,applyTDOM,
  applyFmaskCloudShadowMask,applyFmaskSnowMask,
  cloudScoreThresh,cloudScorePctl,contractPixels,dilatePixels,
  correctIllumination,correctScale,
  exportComposites,outputName,exportPathRoot,crs,transform,scale);

//Separate into scenes and composites for subsequent analysis
var scenes = lsAndTs[0];
var composites = lsAndTs[1];

////////////////////////////////////////////////////////////
//Landtrendr code
var indexDirList = ee.List(indexList).zip(ee.List(ltDirection)).getInfo();

//Iterate across index and direction list
var outputCollection;
var outputStack;
indexDirList.map(function(indexDir){
  var indexName = indexDir[0];
  var distDir = indexDir[1];
  print(indexName,distDir);
  var tsIndex = composites.select([indexName]);
  
  //Run master LT wrapper
  //Returns the raw, heuristic output, and fitted collection
  var ltOutputs = dLib.landtrendrWrapper(tsIndex,startYear+timebuffer,endYear-timebuffer,indexName,distDir,run_params,distParams,mmu);
  
  var ltRaw = ltOutputs[0];
  var ltHeuristic = ltOutputs[1];
  var ltAnnualFitted = ltOutputs[2];
  
  //Stack the heuristic output and stack each image
  //in fitted collection using join
  if(outputCollection === undefined){
    outputCollection = ltAnnualFitted;
    outputStack = ltHeuristic;
  }else{
    outputCollection = getImageLib.joinCollections(outputCollection,ltAnnualFitted,false);
    outputStack = outputStack.addBands(ltHeuristic);
    
  }
  
});
Map.addLayer(outputCollection,{},'LT Fitted IndexNames',false);
Map.addLayer(outputStack.select([0]),{'min':startYear,'max':endYear,'palette':'FF0,F00'},indexList[0] + ' LT Change Year',false);
  
//Export each fitted year
var years = ee.List.sequence(startYear+timebuffer,endYear-timebuffer).getInfo();

  years.map(function(year){
    var ltYr = ee.Image(outputCollection.filter(ee.Filter.calendarRange(year,year,'year')).first())
    .multiply(10000).int16()
    .set('bandsUsed',indexListString)
    .set('system:time_start',ee.Date.fromYMD(year,6,1).millis());
 
  var exportName = outputName + year.toString();
    var exportPath = exportPathRoot + '/'+ exportName;
    
    // getImageLib.exportToAssetWrapper(ltYr,exportName,exportPath,'mean',
      // studyArea,null,crs,transform);
  });

File: examples/LANDTRENDRWrapper.py
"""
This script demonstrates how to run the LANDTRENDR temporal segmentation algorithm and visualize the outputs using Python visualization tools. The script acquires Landsat data, processes it, runs the LANDTRENDR algorithm, and optionally exports the results.
Modules:
    - geeViz.getImagesLib as gil
    - geeViz.changeDetectionLib as cdl
    - geeViz.taskManagerLib as tml
User Parameters:
    - studyArea: The area of interest, can be a featureCollection, feature, or geometry.
    - startJulian: Starting Julian date for seasonal constraints.
    - endJulian: Ending Julian date for seasonal constraints.
    - startYear: Starting year for analysis.
    - endYear: Ending year for analysis.
    - indexName: The band or index to use (e.g., NBR, NDMI, NDVI).
    - howManyToPull: Number of significant loss/gain segments to include.
    - lossMagThresh: Threshold for loss magnitude.
    - lossSlopeThresh: Threshold for loss slope.
    - gainMagThresh: Threshold for gain magnitude.
    - gainSlopeThresh: Threshold for gain slope.
    - slowLossDurationThresh: Threshold for slow loss duration.
    - chooseWhichLoss: Criteria for selecting loss segments.
    - chooseWhichGain: Criteria for selecting gain segments.
    - run_params: Parameters for the LANDTRENDR algorithm.
    - addToMap: Whether to add change outputs to the map.
    - exportLTLossGain: Whether to export LANDTRENDR loss/gain outputs.
    - exportLTVertexArray: Whether to export raw LANDTRENDR vertex array.
    - outputName: Name for the export.
    - exportPathRoot: Path for exporting composites.
    - crs: Coordinate reference system.
    - transform: Transform parameters if scale is None.
    - scale: Scale if transform is None.
Functions:
    - gil.getComposite: Gets the composite image for the specified parameters.
    - gil.getProcessedLandsatScenes: Gets processed Landsat scenes.
    - gil.fillEmptyCollections: Fills empty collections with a dummy image.
    - cdl.simpleLANDTRENDR: Runs the LANDTRENDR algorithm and performs change detection.
    - gil.exportToAssetWrapper: Exports images to an asset.
    - tml.trackTasks2: Tracks export tasks.
Workflow:
    1. Define user parameters.
    2. Acquire and process Landsat data.
    3. Run the LANDTRENDR algorithm.
    4. Optionally export the results.
    5. Visualize the study area and results on the map.
    
   Copyright 2025 Ian Housman

   Licensed under the Apache License, Version 2.0 (the "License");
   you may not use this file except in compliance with the License.
   You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License.
"""

# Example of how to run the LANDTRENDR temporal segmentation algorithm and view outputs using the Python visualization tools
# LANDTRENDR original paper: https://www.sciencedirect.com/science/article/pii/S0034425710002245
# LANDTRENDR in GEE paper: https://www.mdpi.com/2072-4292/10/5/691
# Acquires Landsat data, runs LANDTRENDR, and then adds some outputs to the viewer
####################################################################################################
import os, sys

sys.path.append(os.getcwd())

# Module imports
import geeViz.getImagesLib as gil
import geeViz.changeDetectionLib as cdl
import geeViz.taskManagerLib as tml

ee = gil.ee
Map = gil.Map
Map.port = 1232
Map.clearMap()
####################################################################################################
# Define user parameters:

    # Get composite
    composite = gil.getComposite(studyArea, startYear, endYear, startJulian, endJulian)
# Specify study area: Study area
# Can be a featureCollection, feature, or geometry
studyArea = gil.testAreas["CA"]

# Update the startJulian and endJulian variables to indicate your seasonal
# constraints. This supports wrapping for tropics and southern hemisphere.
# If using wrapping and the majority of the days occur in the second year, the system:time_start will default
# to June 1 of that year.Otherwise, all system:time_starts will default to June 1 of the given year
# startJulian: Starting Julian date
# endJulian: Ending Julian date
startJulian = 152
endJulian = 273

# Specify start and end years for all analyses
# More than a 3 year span should be provided for time series methods to work
# well. If using Fmask as the cloud/cloud shadow masking method, or providing
# pre-computed stats for cloudScore and TDOM, this does not
# matter
startYear = 1990
endYear = 2024


# Choose band or index
# NBR, NDMI, and NDVI tend to work best
# Other good options are wetness and tcAngleBG
indexName = "NBR"

# How many significant loss and/or gain segments to include
# Do not make less than 1
# If you only want the first loss and/or gain, choose 1
# Generally any past 2 are noise
howManyToPull = 1

# Parameters to identify suitable LANDTRENDR segments

# Thresholds to identify loss in vegetation
# Any segment that has a change magnitude or slope less than both of these thresholds is omitted
lossMagThresh = -0.15
lossSlopeThresh = -0.05


# Thresholds to identify gain in vegetation
# Any segment that has a change magnitude or slope greater than both of these thresholds is omitted
gainMagThresh = 0.1
gainSlopeThresh = 0.05

# Number of years of duration to separate between slow and fast loss (>= this number will be called slow loss)
slowLossDurationThresh = 3

# Which segment to show change from
# Choose from: 'newest','oldest','largest','smallest','steepest','mostGradual','shortest','longest'
chooseWhichLoss = "largest"
chooseWhichGain = "largest"

# LandTrendr Params
# run_params ={
#   'timeSeries': (ImageCollection) Yearly time-series from which to extract breakpoints. The first band is used to find breakpoints, and all subsequent bands are fitted using those breakpoints.
#   'maxSegments':            6,\ (Integer) Maximum number of segments to be fitted on the time series.
#   'spikeThreshold':         0.9,\ (Float, default: 0.9) Threshold for damping the spikes (1.0 means no damping).
#   'vertexCountOvershoot':   3,\(Integer, default: 3) The initial model can overshoot the maxSegments + 1 vertices by this amount. Later, it will be pruned down to maxSegments + 1.
#   'preventOneYearRecovery': False,\(Boolean, default: False): Prevent segments that represent one year recoveries.
#   'recoveryThreshold':      0.25,\(Float, default: 0.25) If a segment has a recovery rate faster than 1/recoveryThreshold (in years), then the segment is disallowed.
#   'pvalThreshold':          0.05,\(Float, default: 0.1) If the p-value of the fitted model exceeds this threshold, then the current model is discarded and another one is fitted using the Levenberg-Marquardt optimizer.
#   'bestModelProportion':    0.75,\(Float, default: 0.75) Allows models with more vertices to be chosen if their p-value is no more than (2 - bestModelProportion) times the p-value of the best model.
#   'minObservationsNeeded':  6\(Integer, default: 6) Min observations needed to perform output fitting.
# };
# Define landtrendr params
run_params = {
    "maxSegments": 9,
    "spikeThreshold": 0.9,
    "vertexCountOvershoot": 3,
    "preventOneYearRecovery": False,
    "recoveryThreshold": 0.25,
    "pvalThreshold": 0.05,
    "bestModelProportion": 0.75,
    "minObservationsNeeded": 6,
}

# Whether to add change outputs to map
addToMap = True

# Export params
# Whether to export LANDTRENDR change detection (loss and gain) outputs
exportLTLossGain = False

# Whether to export LandTrendr vertex array raw output
exportLTVertexArray = False

# Set up Names for the export
outputName = "LT_Test"

# Provide location composites will be exported to
# This should be an asset imageCollection
exportPathRoot = "users/username/someCollection"

# CRS- must be provided.
# Common crs codes: Web mercator is EPSG:4326, USGS Albers is EPSG:5070,
# WGS84 UTM N hemisphere is EPSG:326+ zone number (zone 12 N would be EPSG:32612) and S hemisphere is EPSG:327+ zone number
crs = "EPSG:5070"

# Specify transform if scale is None and snapping to known grid is needed
transform = [30, 0, -2361915.0, 0, -30, 3177735.0]

# Specify scale if transform is None
scale = None
####################################################################################################
# End user parameters
####################################################################################################
####################################################################################################
####################################################################################################
# Start function calls
####################################################################################################
hansen = ee.Image("UMD/hansen/global_forest_change_2023_v1_11").select(["lossyear"]).add(2000).int16()
hansen = hansen.updateMask(hansen.neq(2000).And(hansen.gte(startYear)).And(hansen.lte(endYear)))
Map.addLayer(
    hansen,
    {"min": startYear, "max": endYear, "palette": cdl.lossYearPalette},
    "Hansen Loss Year",
    False,
)

####################################################################################################
# Call on master wrapper function to get Landat scenes and composites
# Important that the range of data values of the composites match the run_params spikeThreshold and recoveryThreshold
# e.g. Reflectance bands that have a scale of 0-1 should have a spikeThreshold around 0.9
# and a recoveryThreshold around 0.25
# If the reflectance values were scaled by 10000, the spikeThreshold would be around 9000
# and a recoveryThreshold around 2500

allImages = gil.getProcessedLandsatScenes(studyArea, startYear, endYear, startJulian, endJulian).select([indexName])
dummyImage = allImages.first()
composites = ee.ImageCollection(
    ee.List.sequence(startYear, endYear).map(lambda yr: gil.fillEmptyCollections(allImages.filter(ee.Filter.calendarRange(yr, yr, "year")), dummyImage).median().set("system:time_start", ee.Date.fromYMD(yr, 6, 1).millis()))
)

# Run LANDTRENDR
# This function handles a lot
# This function will ensure the band/index will be flipped so a decrease in veg/moisture goes up
# It will then flip it back and return only the image array of the vertex values and the rmse
# It will then run a simple change detection over the resulting LT output
ltOutputs = cdl.simpleLANDTRENDR(
    composites,
    startYear,
    endYear,
    indexName,
    run_params,
    lossMagThresh,
    lossSlopeThresh,
    gainMagThresh,
    gainSlopeThresh,
    slowLossDurationThresh,
    chooseWhichLoss,
    chooseWhichGain,
    addToMap,
    howManyToPull,
    multBy=10000,
)


if exportLTLossGain:
    lossGainStack = ltOutputs[1]
    # Export  stack
    exportName = f"{outputName}_LT_LossGain_Stack_{indexName}_{startYear}_{endYear}_{startJulian}_{endJulian}"
    exportPath = exportPathRoot + "/" + exportName

    lossGainStack = lossGainStack.set(
        {
            "startYear": startYear,
            "endYear": endYear,
            "startJulian": startJulian,
            "endJulian": endJulian,
            "band": indexName,
        }
    )
    lossGainStack = lossGainStack.set(run_params)

    # Set up proper resampling for each band
    # Be sure to change if the band names for the exported image change
    pyrObj = {"_yr_": "mode", "_dur_": "mode", "_mag_": "mean", "_slope_": "mean"}
    possible = ["loss", "gain"]
    how_many_list = ee.List.sequence(1, howManyToPull).getInfo()
    outObj = {}
    for p in possible:
        for key in pyrObj.keys():
            for i in how_many_list:
                i = int(i)
                kt = indexName + "_LT_" + p + key + str(i)
                outObj[kt] = pyrObj[key]

    # print(outObj)
    # Export output
    gil.exportToAssetWrapper(lossGainStack, exportName, exportPath, outObj, studyArea, scale, crs, transform)


# Export raw LandTrendr array image
if exportLTVertexArray:
    rawLTForExport = ltOutputs[0]
    # Map.addLayer(rawLTForExport,{},'Raw LT For Export {}'.format(indexName),False)

    rawLTForExport = rawLTForExport.set(
        {
            "startYear": startYear,
            "endYear": endYear,
            "startJulian": startJulian,
            "endJulian": endJulian,
            "band": indexName,
        }
    )
    rawLTForExport = rawLTForExport.set(run_params)
    exportName = "{}_LT_Raw_{}_{}_{}_{}_{}".format(outputName, indexName, startYear, endYear, startJulian, endJulian)
    exportPath = exportPathRoot + "/" + exportName
    gil.exportToAssetWrapper(
        rawLTForExport,
        exportName,
        exportPath,
        {".default": "sample"},
        studyArea,
        scale,
        crs,
        transform,
    )
    # Reverse for modeling
    # decompressedC = cdl.simpleLTFit(rawLTForExport,startYear,endYear,indexName,True,run_params['maxSegments'])
    # Map.addLayer(decompressedC,{},'Decompressed LT Output {}'.format(indexName),False)

####################################################################################################
# Load the study region
Map.addLayer(studyArea, {"strokeColor": "0000FF"}, "Study Area", False)
Map.centerObject(studyArea)
####################################################################################################
####################################################################################################
# View map
Map.turnOnInspector()
Map.view()
####################################################################################################
####################################################################################################
# If exporting LT stack, track the exports
if exportLTLossGain or exportLTVertexArray:
    tml.trackTasks2()
